pymc3/math.py

```diff
@@ -18,10 +18,84 @@
 import scipy.sparse
 from scipy.linalg import block_diag as scipy_block_diag
 from pymc3.theanof import floatX, largest_common_dtype, ix_
+from functools import reduce, partial
 
 # pylint: enable=unused-import
 
 
+def kronecker(*Ks):
+    """Return the Kronecker product of arguments:
+            K_1 \otimes K_2 \otimes ... \otimes K_D
+
+    Parameters
+    ----------
+    Ks: 2D array-like
+    """
+    return reduce(tt.slinalg.kron, Ks)
+
+
+def cartesian(*arrays):
+    """Makes the Cartesian product of arrays.
+
+    Parameters
+    ----------
+    arrays: 1D array-like
+            1D arrays where earlier arrays loop more slowly than later ones
+    """
+    N = len(arrays)
+    return np.stack(np.meshgrid(*arrays, indexing='ij'), -1).reshape(-1, N)
+
+
+def kron_matrix_op(krons, m, op):
```

Comment:
I think this is really nice

Reply:
Thanks! Getting it working in Theano was much tricker than Numpy. If there are more efficient ways than `scan` and the nested functions that would be great to see.

Reply:
I profiled `model.logp` and the gradient, and the scan here is the slowest part (~50% of computation time).  I'm ok with saving this optimization for another PR though.

Reply:
Oh thanks for looking into that! Would certainly like to check that out in some future PR.

Reply:
[heres where I was looking at that](https://gist.github.com/bwengals/bfb0a97d2cf65a5eb4261298e8e2d5d0), down towards the bottom.  You can run the theano profiler on pymc3 code.  You can see that scan at the top.

Reply:
Hi, I am trying to use KroneckerNormal directly (not through the GP interface) but unfortunately it seems prohibitively slow for my application (though it could be something I'm doing wrong, not a theano expert by any means). My profiling also implicated scan for ~50% of the computation time, but overall it was much slower than @bwengals example (I think ~267 seconds for 1000 evaluations of the likelihood). I tried it with data that is 1000 by 2, flattened; I define one covariance matrix (1000 by 1000) using a squared exponential covariance function that operates on a 5-dimensional input, and the second (2 by 2) is unconstrained and given an LKJChol prior. I'm trying to do MAP estimation and it takes a very long time. I was doing it through MatrixNormal before and it was pretty fast, but I wanted to be able to put additive noise (and was happy to find this is implemented here!). Unfortunately I can't share my code right now, but I could concoct a toy example later to share if that is helpful. Thanks for any input you all might have -- I simply am curious whether this is inherent to current implementation due to theano stuff, whether I might be doing something wrong, or whether the nature of what I'm doing is sufficiently different from the single-input GP examples that it's slowing it down.

Reply:
@natalieklein the easiest way to get some support on this is probably to post it on our [discourse site](discourse.pymc.io) along with a complete example that demonstrates the issue.

Reply:
Thanks! If anyone is interested, here's the discourse link: https://discourse.pymc.io/t/kroneckernormal-speed/1497